### ISSUE TITLE
[Form] How to disable form item input warnings on console

### ISSUE THREAD

**Issue by bichenkk** [ant-design/ant-design#9412]
### What problem does this feature solve?
When the input of Form Item is invalid, it logs on console.
Example: async-validator: ["email is not a valid email"]

### What does the proposed API look like?
Can we have any flag to disable the warnings on console?

**Comment by MiroslavPetrik** [comment]
@yesmeck what about test environment? The warnings pollute the terminal output.

**Comment by yesmeck** [comment]
@MiroslavPetrik We use async-validator under the hood, you can file a new issue to https://github.com/yiminghe/async-validator/issues

**Comment by FrankenSteinxD** [comment]
@yesmeck  We should be able to disable this even on development.

**Comment by simon-paris** [comment]
I've also made a PR to fix this in async-validator, although the issue was raised over there in march and nobody replied, so idk if they'll reply to me.

@bichenkk @MiroslavPetrik @FrankenSteinxD If you're using webpack, here's a loader that'll remove the warning

```
module.exports = function stupidWarningFixer(content) {
	return content.replace(
`// don't print warning message when in production env or node runtime
if (process.env.NODE_ENV !== 'production' && typeof window !== 'undefined' && typeof document !== 'undefined') {
  warning = function warning(type, errors) {
    if (typeof console !== 'undefined' && console.warn) {
      if (errors.every(function (e) {
        return typeof e === 'string';
      })) {
        console.warn(type, errors);
      }
    }
  };
}`, '\n\n\n\n\n\n\n\n\n\n\n');
};
```
```
{
	test: /\.js$/,
	loader: path.join(__dirname, './stupidWarningFixer.js'),
	include: path.join(process.cwd(), './node_modules/async-validator/')
},
```

**Comment by cdagli** [comment]
An option added for async-validator: https://github.com/yiminghe/async-validator/issues/92#issuecomment-506602112

**Comment by oyalhi** [comment]
@cdagli  yes it seems so. However, how do we enable/disable the `suppressWarning` option using Antd is still not clear. Any pointers?

**Comment by exlame** [comment]
You can specify `async-validator` options like this : 
```
form.validateFields(
  { suppressWarning: true },
  (err, values) => {
        ...
  },
);
```
Note that this will not remove the ` console.error()` that is thrown by  [rc-form](https://github.com/react-component) (https://github.com/react-component/form/blob/32721598fb40be845105964004af87fe2a06fcd1/src/createBaseForm.js#L640), AND it will not remove the warning thrown by async-validator on `onChange` events because rc-form doesn't give the possibility to add custom options on this event: https://github.com/react-component/form/issues/313

I created a PR to remove the error thrown by [rc-form](https://github.com/react-component)
https://github.com/react-component/form/pull/366

**Comment by oyalhi** [comment]
`validateFields` takes in `ValidateFieldsOptions` and I don't see `suppressWarning` in the options.  TypeScript also gives an error.

**Comment by exlame** [comment]
`validateFields()` pass automatically all options to `async-validator.validate()`  but the TypeScript interface doesn't reflect `suppressWarning` ATM

**Comment by ldsenow** [comment]
I believe the suppressWarning option should be default to disabled. It doesnt make much sense to enable it by default

**Comment by phifa** [comment]
How can I use suppressWarning when using the custom validator like so? @exlame 

```
        rules={[
              {
                required: true,
                message: 'Please input your password!',
              },
              () => ({
                validator(rule, value) {
                  if (!value || value.length < 6) {
                    return Promise.reject(
                      new Error(
                        'Password has to contain at least 6 characters.',
                      ),
                    );
                  }
                  return Promise.resolve();
                },
              }),
              () => ({
                validator() {
                  if ('password' in formErrors) {
                    return Promise.reject(formErrors.password);
                  }
                  return Promise.resolve();
                },
              }),
            ]}
```

**Comment by marcellomontemagno** [comment]
Could this issue be possibly re-opened? At the moment the only viable option seems to have tons of warning logged while running our test suit.

Passing an extra argument to `validateFields` seems to be a partial solution as warnings would still logged on change.

Mocking `console.warn` is a dangerous option as you will miss warning you might care about, I'm personally not considering it.

**Comment by sauldeleon** [comment]
Hi
@marcellomontemagno this is not a solution for this problem, but what I have done is to mock only console.warn on desired messages.

this is the code I set in a setupJest.ts file, before all tests start to run

```ts
const filteredWarnMessages: string[] = [
  'async-validator:',
  'componentWillReceiveProps has been renamed',
]
const privateWarnLog = console.warn
jest.spyOn(console, 'warn').mockImplementation((msg: string, ...args: unknown[]) => {
  filteredWarnMessages.some(message => msg.includes(message)) ? jest.fn() : privateWarnLog(msg, ...args)
})

// List of message errors we want to filter from error logs in tests
const filteredErrorMessages: string[] = []
const privateErrorLog = console.error
jest.spyOn(console, 'error').mockImplementation((msg: string, ...args: unknown[]) => {
  filteredErrorMessages.some(message => msg.includes(message)) ? jest.fn() : privateErrorLog(msg, ...args)
})
```

With this, I select which messages I want to hide, in case they are related with external libraries where I don't have control

**Comment by ShivamJoker** [comment]
Any updates on this ?
To disable it in all components ?

**Comment by CANHBK** [comment]
I also think this should be disabled by default

**Comment by qwerdee** [comment]
Agreed, this should be disabled by default!

**Comment by AymenBenAmor** [comment]
Agreed, this should be disabled by default!

**Comment by gtb104** [comment]
I tried @sauldeleon's workaround, but couldn't get it to work.
```javascript
beforeAll(() => {
  const filters = ['async-validator:'];
  const _warn = console.warn;
  jest.spyOn(console, 'warn').mockImplementation(
    (msg, ...args) => {
      filters.some((filter) => msg.includes(filter)) ? jest.fn() : _warn(msg, ...args);
    }
  );
});
```
The validators messages aren't being filtered out, ALL console.warn messages are being removed.

**Comment by sauldeleon** [comment]
Hey @gtb104 

I have this piece of code in my `setupTests.ts` file within the root folder. Have tested in 2 projects and working for now

```tsx
// List of message warnings we want to filter from warning logs in tests
const filteredWarnMessages: string[] = [
  'React.createFactory()',
  'async-validator:',
  'componentWillReceiveProps has been renamed',
  'value provided is not in a recognized RFC2822 or ISO format',
]
const filteredWarnVariables: string[] = []
const privateWarnLog = console.warn
jest.spyOn(console, 'warn').mockImplementation((msg: string, ...args: unknown[]) => {
  filteredWarnMessages.some(message => msg.includes(message)) ||
  filteredWarnVariables.some(variable => variable === args[0])
    ? jest.fn()
    : privateWarnLog(msg, ...args)
})

// List of message errors we want to filter from error logs in tests
const filteredErrorMessages: string[] = [
  'when doing server-side rendering',
  'If you want to write it to the DOM, pass a string instead',
]

const filteredErrorVariables: string[] = ['warnKey']
const privateErrorLog = console.error
jest.spyOn(console, 'error').mockImplementation((msg: string, ...args: unknown[]) => {
  filteredErrorMessages.some(message => msg.includes(message)) ||
  filteredErrorVariables.some(variable => variable === args[0])
    ? jest.fn()
    : privateErrorLog(msg, ...args)
})
```

Honestly, I am not doing anything else. If you put that code in the setupTest file of your project, does it get filter correctly the messages?

When I remove some lines from the filter

![image](https://user-images.githubusercontent.com/2400723/130798044-6bf0e31c-ba7f-41a8-abf2-14edb24caaa3.png)

And adding it again, removes the message!

**Comment by gtb104** [comment]
And you get other warnings?

What I wrote, besides being plain JS, is basically the same. I wonder why it seems to be whacking all console.warn messages.

**Comment by gtb104** [comment]
I'm wondering if the fact that we used CRA to set this project up, and that we use react-app-rewired to run the tests is causing issues for me.

**Comment by gtb104** [comment]
This is the only way I could get it to work in our env
```jsx
const filters = ['async-validator:'];
const _warn = console.warn;
console.warn = function(msg, ...args) {
  filters.some((filter) => msg.includes(filter)) ? jest.fn() : _warn.apply(console, [msg, ...args]);
};
```

**Comment by ArturBaybulatov** [comment]
This issue should be reopened, since the initial problem is not solved for non-prod environments.

**Comment by biruelrick** [comment]
> There's a couple of ways to disable async-validator's warnings within the Ant app:
> 
> ```js
> import Schema from 'async-validator';
> Schema.warning = function(){};
> ```
> 
> Other options are described here:
> 
> https://github.com/yiminghe/async-validator#how-to-avoid-global-warning

If the above solution not work (>4.20.0), use the following code in App.js after last import instruction...

global.ASYNC_VALIDATOR_NO_WARNING = 1;

**Comment by IanTorquato** [comment]
> global.ASYNC_VALIDATOR_NO_WARNING = 1;
Thank you @biruelrick 😉

Using Typescript + CRA I solved it by putting the same code above in the `src/setupTests.ts` file.
`(global as any).ASYNC_VALIDATOR_NO_WARNING = 1;`

https://create-react-app.dev/docs/running-tests/#initializing-test-environment
> If your app uses a browser API that you need to mock in your tests or if you need a global setup before running your tests, add a src/setupTests.js to your project. It will be automatically executed before running your tests.
